moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.kt

```diff
@@ -0,0 +1,365 @@
+      }
+    } else {
+      null
+    }
+
+    toAdapter?.bind(moshi, this)
+    fromAdapter?.bind(moshi, this)
+
+    return object : JsonAdapter<Any>() {
+      override fun toJson(writer: JsonWriter, value: Any?) {
+        if (toAdapter == null) {
+          requireNotNull(delegate).toJson(writer, value)
+        } else if (!toAdapter.nullable && value == null) {
+          writer.nullValue()
+        } else {
+          try {
+            toAdapter.toJson(moshi, writer, value)
+          } catch (e: InvocationTargetException) {
+            val cause = e.cause
+            if (cause is IOException) throw cause
+            throw JsonDataException("$cause at ${writer.path}", cause)
+          }
+        }
+      }
+
+      override fun fromJson(reader: JsonReader): Any? {
+        return if (fromAdapter == null) {
+          requireNotNull(delegate).fromJson(reader)
+        } else if (!fromAdapter.nullable && reader.peek() == JsonReader.Token.NULL) {
+          reader.nextNull<Any>()
+          null
+        } else {
+          try {
+            fromAdapter.fromJson(moshi, reader)
+          } catch (e: InvocationTargetException) {
+            val cause = e.cause
+            if (cause is IOException) throw cause
+            throw JsonDataException("$cause at ${reader.path}", cause)
+          }
+        }
+      }
+
+      override fun toString() = "JsonAdapter$annotations($type)"
+    }
+  }
+
+  companion object {
+    fun get(adapter: Any): AdapterMethodsFactory {
+      val toAdapters = mutableListOf<AdapterMethod>()
+      val fromAdapters = mutableListOf<AdapterMethod>()
+
+      var clazz = adapter.javaClass
+      while (clazz != Any::class.java) {
+        clazz.declaredMethods.forEach { declaredMethod ->
+          if (declaredMethod.isAnnotationPresent(ToJson::class.java)) {
+            val toAdapter = toAdapter(adapter, declaredMethod)
+            val conflicting = get(toAdapters, toAdapter.type, toAdapter.annotations)
+            require(conflicting == null) {
+              "Conflicting @ToJson methods:\n" +
+                "    " +
+                requireNotNull(conflicting).method +
+                "\n" +
+                "    " +
+                toAdapter.method
+            }
+            toAdapters.add(toAdapter)
+          }
+          if (declaredMethod.isAnnotationPresent(FromJson::class.java)) {
+            val fromAdapter = fromAdapter(adapter, declaredMethod)
+            val conflicting = get(fromAdapters, fromAdapter.type, fromAdapter.annotations)
+            require(conflicting == null) {
+              "Conflicting @FromJson methods:\n" +
+                "    " +
+                requireNotNull(conflicting).method +
+                "\n" +
+                "    " +
+                fromAdapter.method
+            }
+            fromAdapters.add(fromAdapter)
+          }
+        }
+        clazz = clazz.superclass
+      }
+      require(!(toAdapters.isEmpty() && fromAdapters.isEmpty())) {
```

Comment:
Simplify this logic for readability (switch to an `||`)

Reply:
https://github.com/square/moshi/pull/1467/commits/1b2274783ed67a9de871bd4a188d31e98fa5c6ba